File: Src/ButterflyNet.Parser/Scopes/TableScope.cs
namespace ButterflyNet.Parser.Scopes {
	public class TableScope : BlockScope {
		public override bool CanNestText { get { return false; } }

		protected override void OpenAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.OpenTable();
		}

		protected override void CloseAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.CloseTable();
		}
	}

	public class TableRowScope : BlockScope {
		public override bool CanNestText { get { return false; } }

		protected override void OpenAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.OpenTableRow();
		}

		protected override void CloseAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.CloseTableRow();
		}
	}

	public class TableRowLineScope : BlockScope {
		public override bool CanNestText { get { return false; } }
		public override bool CloseOnSingleLineBreak { get { return true; } }
		public override bool ManuallyClosing { get { return false; } }

		protected override void OpenAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.OpenTableRowLine();
		}

		protected override void CloseAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.CloseTableRowLine();
		}
	}

	public class TableCellScope : BlockScope {
		protected override void OpenAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.OpenTableCell();
		}

		protected override void CloseAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.CloseTableCell();
		}
	}

	public class TableHeaderScope : BlockScope {
		protected override void OpenAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.OpenTableHeader();
		}

		protected override void CloseAndAnalyze(ButterflyAnalyzer analyzer) {
			analyzer.CloseTableHeader();
		}
	}
}

File: src/js/src/scopes/TableScope.js
function TableRowLineScope() {
	this.type = ScopeTypeCache.tableRowLine;
	this.level = ScopeLevel.block;
	this.open = function(analyzer) {
		analyzer.openTableRowLine();
	};
	this.close = function(analyzer) {
		analyzer.closeTableRowLine();
	};
}

extend(Scope, TableRowLineScope);

function TableRowScope() {
	this.type = ScopeTypeCache.tableRow;
	this.level = ScopeLevel.block;
	this.open = function(analyzer) {
		analyzer.openTableRow();
	};
	this.close = function(analyzer) {
		analyzer.closeTableRow();
	};
}

extend(Scope, TableRowScope);

function TableScope() {
	this.type = ScopeTypeCache.table;
	this.level = ScopeLevel.block;
	this.open = function(analyzer) {
		analyzer.openTable();
	};
	this.close = function(analyzer) {
		analyzer.closeTable();
	};
}

extend(Scope, TableScope);

function TableCellScope() {
	this.type = ScopeTypeCache.tableCell;
	this.level = ScopeLevel.block;
	this.canNestText = true;
	this.open = function(analyzer) {
		analyzer.openTableCell();
	};
	this.close = function(analyzer) {
		analyzer.closeTableCell();
	};
}

extend(Scope, TableCellScope);

function TableHeaderScope() {
	this.type = ScopeTypeCache.tableHeader;
	this.level = ScopeLevel.block;
	this.canNestText = true;
	this.open = function(analyzer) {
		analyzer.openTableHeader();
	};
	this.close = function(analyzer) {
		analyzer.closeTableHeader();
	};
}

extend(Scope, TableHeaderScope);